euiAnimator header

break-core opened the pull request:
This adds in eui's Animator, a core part of the eui system that derives from AnimTransform on ui2d.

It should be noted that the eui version of Animator utilizes things that the ui2d version doesn't, such as sead's Random classes or LayoutEx (which is not yet implemented in this repository, but probably will be soon).

Information about the PlayType enum was pulled from MK8DX, as well as things like file paths.

break-core committed:
Implement euiAnimator header

ThePixelGamer commented on include/sys/euiAnimator.h:
```diff
@@ -0,0 +1,45 @@
+#include <nn/ui2d/ui2d_Group.h>
+#include <nn/ui2d/ui2d_Pane.h>
+
+#include <eui/lytex/euiLayoutEx.h>
```
Don't include files that don't exist or create a dummy version of that header with a fwd decl, also the include isn't needed here just do a fwd decl of the class

ThePixelGamer commented on include/sys/euiAnimator.h:
```diff
@@ -0,0 +1,45 @@
+
+namespace eui {
+
+class Animator : public AnimTransformBasic {
```
Is AnimTransformBasic from ui2d? If so you're missing the namespace or class

break-core replied:
AnimTransformBasic is part of ui2d_Animation. You're right, I should probably fix that.

ThePixelGamer commented on include/sys/euiAnimator.h:
```diff
@@ -0,0 +1,45 @@
+
+class Animator : public AnimTransformBasic {
+public:
+    NN_RUNTIME_TYPEINFO_BASE();
```
This is probably not right, I would check the functions in the binary and see if they reference the parents' runtime vars (in which case use the other macro)

break-core replied:
I've changed the implementation of this to now use NN_RUNTIME_TYPEINFO with AnimTransformBasic. Should be approximate to what the original headers used.

ThePixelGamer commented on include/sys/euiAnimator.h:
```diff
@@ -0,0 +1,45 @@
+    virtual void SetEnabled(bool bEnabled);
+
+    virtual void SetupBasic(const AnimResource &animRes, LayoutEx *pLayoutEx, bool enabled);
+    virtual void SetupWithGroup(const AnimResource &animRes, LayoutEx *pLayoutEx, Group *pGroup, bool enabled);
```
Missing ui2d namespace

ThePixelGamer commented on include/sys/euiAnimator.h:
Can you run clang format? It might not catch it but pointers/references should be left handed

break-core committed:
Fix most of the issues

- changed RTTI to be NN_RUNTIME_TYPEINFO(AnimTransformBasic)
- add namespaces properly to types
- forward decl. classes that we don't have yet in nnheaders or eui.

break-core committed:
fix namespace issue with AnimTransformBasic

ThePixelGamer commented on include/sys/euiAnimator.h:
```diff
@@ -0,0 +1,49 @@
+    virtual void PlayRandom(PlayType type, f32 speed);
+    virtual void SetEnabled(bool bEnabled);
+
+    virtual void SetupBasic(const nn::ui2d::AnimResource &animRes, eui::LayoutEx *pLayoutEx, bool enabled);
```
leaving out the eui namespace is fine seeing how we're in it in this class

ThePixelGamer commented on include/sys/euiAnimator.h:
```diff
@@ -0,0 +1,49 @@
+namespace nn {
+namespace ui2d {
+    class Group;
+    class AnimTransformBasic;
```
AnimTransformBasic needs to be included for this class as it's part of the inheritance tree

break-core replied:
Sorry if I don't know what I'm talking about, but what would you mean by this?

break-core replied:
Never mind, I think I figured it out. So I would assume you want me to include the header that AnimTransformBasic comes?

ThePixelGamer commented on include/sys/euiAnimator.h:
```diff
@@ -0,0 +1,49 @@
+#pragma once
+
+#include <nn/font/font_Util.h>
+#include <nn/ui2d/ui2d_Pane.h>
```
Pane can be left out I believe

break-core replied:
I have to keep Pane because there is one specific function that uses Pane called `SetupWithPane`. Very specific use-case but it is required for eui's Animator.

ThePixelGamer replied:
The cpp file (where that function will be implemented) can have the include, this header doesn't seem to need it from what I can tell (usually you only want to add a header include when it's needed for a function implementation or part of the class size calculations)

break-core committed:
clang-format pass, AnimTransformBasic include fix

ThePixelGamer commented on include/sys/euiAnimator.h:
```diff
@@ -1,15 +1,16 @@
-}
+class Group;
+class AnimTransformBasic;
+class AnimResource;
```
AnimTransformBasic and AnimResource can be removed seeing how they're defined in the required header (ui2d_Animation), Pane should be added to these forward decls imo

break-core replied:
Alright. I'll forward decl. Pane just because and remove AnimTransformBasic and AnimResource.

break-core committed:
Un-forward and forward declare some classes.

- Pane is now forward declared (by request of maintainer)
- ui2d_Animation included

break-core committed:
clang-format check changes

HOPEFULLY will pass clang-format according to it's fails listed.

MonsterDruide1 commented on include/sys/euiAnimator.h:
```diff
@@ -0,0 +1,49 @@
+public:
+    NN_RUNTIME_TYPEINFO(AnimTransformBasic);
+
+    enum PlayType { PLAYTYPE_ONESHOT, PLAYTYPE_LOOP, PLAYTYPE_ROUNDTRIP };
```
Should be `enum class` to avoid it being replaced by the base type in function symbols.

Take this with a grain of salt, I don't really know if the above statement is true - `enum class` is preferred by the contribution guidelines either way, if there is no indication that this should be a normal `enum` instead of `enum class`.

break-core replied:
I'm 100% sure that `enum class` was not used at all.

MonsterDruide1 replied:
How do you know that / how would you tell apart whether `enum class` or `enum` was used?

break-core replied:
`enum` seems to be the most likely implementation for this case. There isn't really a way to tell, I just simply used `enum` as it seemed like Nintendo.

I could change it to `enum class` if it really bugs you or any of the other maintainers.

ThePixelGamer replied:
there are cases where enum is more likely (in this case maybe there was a print or string for these values), also I'm pretty sure enum still produces the proper symbol

leoetlino replied:
Unless we have debug info or asserts or whatever, I don't think it's possible to tell if they used a regular enum or an enum class. Both types of enums are mangled the exact same way.

In this case it's more idiomatic to use an enum class - we don't rely on implicit conversion to int and there is no good reason for the enumerators to leak into their surrounding scope. The PLAYTYPE_ prefix and SCREAMING would be unnecessary with an enum class.

In general what the original Nintendo code does is merely a guideline rather than something to be followed rigidly (as long as things produce the same codegen). We frequently deviate from what we think N did if it makes the code clearer or if we can modernise the code without introducing codegen differences. In this specific case we don't even have evidence that they used an old style enum so I'd recommend changing this to an enum class.

break-core replied:
Alright, I'll change it to an `enum class` because I don't feel like arguing about this much more.

break-core committed:
Change PlayType from `enum` to `enum class`.

MonsterDruide1 commented on include/sys/euiAnimator.h:
```diff
@@ -0,0 +1,49 @@
+    virtual void StopCurrent();
+    virtual void Synchronize(const Animator& animator);
+    virtual void UpdateFrame(f32 progress_frame);
+}
```
```suggestion
};
```
This should even be a compiler error...

break-core committed:
Add semicolon at brace end

(very dumb mistake I made woops lol)

Co-authored-by: MonsterDruide1 <5958456@gmail.com>
